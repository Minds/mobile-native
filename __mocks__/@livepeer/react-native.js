// mock @livepeer/react-native
export const LivepeerConfig = {
  setConfig: jest.fn(),
  getConfig: jest.fn(),
  get: jest.fn(),
  set: jest.fn(),
};

export const createReactClient = jest.fn();
export const createLivepeerClient = jest.fn();
export const Player = jest.fn();
