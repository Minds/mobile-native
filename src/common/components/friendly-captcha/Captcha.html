<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Captch</title>
  </head>
  <body style="background-color: transparent">
    <div id="captcha"></div>
  </body>
  <script
    type="module"
    src="https://unpkg.com/friendly-challenge@0.9.1/widget.module.min.js"
  ></script>
  <script
    nomodule
    src="https://unpkg.com/friendly-challenge@0.9.1/widget.min.js"
  ></script>
  <script>
    const postMessage = window.ReactNativeWebView
      ? window.ReactNativeWebView.postMessage
      : console.log;

    document.addEventListener('DOMContentLoaded', function (event) {
      try {
        new friendlyChallenge.WidgetInstance(
          document.getElementById('captcha'),
          {
            startMode: 'auto',
            sitekey: 'FCMNS7BU7R8316DQ',
            // puzzleEndpoint:
            //   'https://feat-friendly-captcha-2272.minds.io/api/v3/friendly-captcha/puzzle',
            doneCallback: solution => {
              postMessage(`SUCCESS: ${solution}`);
            },
            errorCallback: error => {
              postMessage(`ERROR: ${error}`);
            },
          },
        );
      } catch (e) {
        postMessage(`ERROR: ${e}`);
        console.error(e);
      }
    });
  </script>
</html>
