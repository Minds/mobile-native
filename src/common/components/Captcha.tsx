import React, { forwardRef, useImperativeHandle, useRef } from 'react';
import { View, Image, StyleSheet, TextInput, Text } from 'react-native';
import { observer, useLocalStore } from 'mobx-react';
import Modal from 'react-native-modal';
import { toJS } from 'mobx';

import api from '../services/api.service';
import type { ApiResponse } from '../services/api.service';
import ThemedStyles from '../../styles/ThemedStyles';
import i18n from '../services/i18n.service';
import { Icon } from 'react-native-elements';
import { DARK_THEME } from '../../styles/Colors';
import i18nService from '../services/i18n.service';
import InputContainer from './InputContainer';

const backgroundPrimary = { backgroundColor: DARK_THEME.primary_background };
const backgroundSecondary = {
  backgroundColor: DARK_THEME.secondary_background,
  borderColor: DARK_THEME.borderColorPrimary,
};

interface CaptchaResponse extends ApiResponse {
  base64_image: string;
  jwt_token: string;
}

type PropsType = {
  onResult: (result: string) => void;
  testID?: string;
};

/**
 * Captcha
 */
const Captcha = observer(
  forwardRef((props: PropsType, ref: any) => {
    const theme = ThemedStyles.style;

    const store = useLocalStore(() => ({
      show: false,
      error: '',
      captchaImage: { uri: '' as string },
      clientText: '' as string,
      jwtToken: '' as string,
      setText(value: string) {
        store.clientText = value;
      },
      setError(value: string) {
        store.error = value;
      },
      hideModal() {
        store.show = false;
      },
      showModal() {
        store.show = true;
        store.clientText = '';
        store.load();
      },
      send() {
        if (!store.clientText) {
          return;
        }
        props.onResult(
          JSON.stringify({
            jwtToken: store.jwtToken,
            clientText: store.clientText,
          }),
        );
      },
      async load() {
        const response: CaptchaResponse = await api.get('api/v2/captcha');
        store.captchaImage.uri = response.base64_image;
        store.jwtToken = response.jwt_token;
      },
    }));

    useImperativeHandle(ref, () => ({
      show: () => {
        store.showModal();
      },
      hide: () => {
        store.hideModal();
      },
    }));

    const src = toJS(store.captchaImage);

    return (
      <Modal
        avoidKeyboard={true}
        onBackdropPress={store.hideModal}
        isVisible={store.show}
        backdropColor={DARK_THEME.secondary_background}
        backdropOpacity={0.9}
        useNativeDriver={true}
        style={[theme.fullWidth, theme.margin0x, theme.justifyEnd]}
        animationInTiming={100}
        animationOutTiming={100}
        animationOut="fadeOut"
        animationIn="fadeIn">
        <View style={[styles.modal, backgroundPrimary]}>
          <View
            style={[theme.paddingHorizontal4x, theme.rowJustifySpaceBetween]}>
            <Text
              onPress={store.hideModal}
              style={[
                theme.fontXL,
                theme.colorWhite,
                theme.paddingVertical4x,
                theme.textCenter,
              ]}>
              {i18nService.t('close')}
            </Text>
            <Text
              style={[
                theme.fontXL,
                theme.colorWhite,
                theme.paddingVertical4x,
                theme.textCenter,
                theme.bold,
              ]}>
              {i18nService.t('verification')}
            </Text>
            <Text
              onPress={store.send}
              style={[
                theme.fontXL,
                theme.paddingVertical4x,
                theme.textCenter,
                theme.colorLink,
              ]}>
              {i18nService.t('verify')}
            </Text>
          </View>
          {store.captchaImage.uri !== '' && (
            <View
              style={[
                theme.rowJustifyStart,
                theme.alignCenter,
                theme.rowJustifyCenter,
                theme.paddingVertical4x,
              ]}>
              <Image source={src} style={styles.image} />
              <Icon
                name="reload"
                type="material-community"
                underlayColor={ThemedStyles.getColor('primary_background')}
                size={45}
                iconStyle={theme.colorIcon}
                onPress={store.load}
              />
            </View>
          )}
          <InputContainer
            labelStyle={theme.colorWhite}
            containerStyle={backgroundSecondary}
            style={theme.colorWhite}
            placeholder={i18n.t('captcha')}
            onChangeText={store.setText}
            onEndEditing={store.send}
            testID="captchInput"
            autofocus
          />
        </View>
      </Modal>
    );
  }),
);

export default Captcha;

const styles = StyleSheet.create({
  modal: {
    borderTopLeftRadius: 20,
    borderTopRightRadius: 20,
    justifyContent: 'center',
    paddingBottom: 50,
    flexDirection: 'column',
    width: '100%',
  },
  image: {
    width: 250,
    height: 100,
  },
});
