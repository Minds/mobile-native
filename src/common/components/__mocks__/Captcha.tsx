import React, { Component } from 'react';
import { View } from 'react-native';

export default class Captcha extends Component<{ onResult: (string) => void }> {
  show() {
    this.props.onResult(
      JSON.stringify({
        jwtToken: 'FAFA',
        clientText: 'some45',
      }),
    );
  }

  hide() {}

  render() {
    return <View />;
  }
}
