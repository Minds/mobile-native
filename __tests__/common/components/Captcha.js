import 'react-native';
import React from 'react';
import Captcha from '../../../src/common/components/Captcha';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const callback = jest.fn();
  const captcha = renderer.create(<Captcha onResult={callback} />).toJSON();
  expect(captcha).toMatchSnapshot();
});
