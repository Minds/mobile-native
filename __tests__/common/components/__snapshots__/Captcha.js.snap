// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Modal
  animationType="none"
  deviceHeight={null}
  deviceWidth={null}
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={false}
  onBackdropPress={[Function]}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
    ]
  }
  swipeThreshold={100}
  transparent={true}
  visible={false}
>
  <View
    accessible={true}
    focusable={true}
    forwardedRef={[Function]}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#1A2025",
        "bottom": 0,
        "height": 1334,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 750,
      }
    }
  />
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      Array [
        Array [
          Object {
            "margin": 37.5,
            "transform": Array [
              Object {
                "translateY": 0,
              },
            ],
          },
          Object {
            "flex": 1,
            "justifyContent": "center",
          },
          Array [
            undefined,
            undefined,
            undefined,
          ],
          Object {
            "margin": 0,
          },
        ],
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      deviceHeight={null}
      deviceWidth={null}
      forwardedRef={[Function]}
      hideModalContentWhileAnimating={false}
      onBackdropPress={[Function]}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
          "margin": 37.5,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeThreshold={100}
    >
      <View
        style={
          Array [
            Object {
              "borderTopLeftRadius": 20,
              "borderTopRightRadius": 20,
              "flexDirection": "column",
              "justifyContent": "center",
              "paddingBottom": 50,
              "width": "100%",
            },
            Object {
              "backgroundColor": "#242A30",
            },
          ]
        }
      >
        <View
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        >
          <Text
            onPress={[Function]}
            style={
              Array [
                undefined,
                undefined,
                undefined,
                undefined,
              ]
            }
          >
            Close
          </Text>
          <Text
            style={
              Array [
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
              ]
            }
          >
            Verification
          </Text>
          <Text
            onPress={[Function]}
            style={
              Array [
                undefined,
                undefined,
                undefined,
                undefined,
              ]
            }
          >
            Verify
          </Text>
        </View>
        <View
          style={
            Array [
              undefined,
              Object {
                "paddingTop": 10,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              Object {
                "backgroundColor": "#1A2025",
                "borderColor": "#404a4e",
              },
            ]
          }
        >
          <View>
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  Object {
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "Roboto",
                        "fontSize": 16,
                        "fontWeight": "600",
                        "marginBottom": 5,
                      },
                      undefined,
                      undefined,
                    ]
                  }
                >
                  Enter the characters above
                </Text>
              </View>
            </View>
            <TextInput
              allowFontScaling={true}
              autoCapitalize="none"
              autofocus={true}
              containerStyle={
                Object {
                  "backgroundColor": "#1A2025",
                  "borderColor": "#404a4e",
                }
              }
              onChangeText={[Function]}
              onEndEditing={[Function]}
              placeholder=""
              placeholderTextColor="#444"
              rejectResponderTermination={true}
              returnKeyType="done"
              style={
                Array [
                  undefined,
                  Array [
                    Object {
                      "borderWidth": 0,
                      "marginBottom": 7,
                      "paddingHorizontal": 0,
                      "paddingVertical": 8,
                    },
                    undefined,
                    undefined,
                    undefined,
                  ],
                ]
              }
              testID="captchInput"
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
